pyro/contrib/tracking/extended_kalman_filter.py

```diff
@@ -0,0 +1,238 @@
+        elif destination_time is None:
+            assert dt is not None
+            destination_time = self._time + dt
+
+        self._mean = self._dynamic_model(self._mean, dt)
+
+        F = self._dynamic_model.jacobian(dt)
+        Q = self._dynamic_model.process_noise_cov(dt)
+        self._cov = F.mm(self._cov).mm(F.t()) + Q
+
+        self._time = destination_time
+
+        self._clear_cached()
+
+    def innovation(self, measurement):
+        '''
+        Compute and return the innovation that a measurement would induce if
+        it were used for an update, but don't actually perform the update.
+        Assumes state and measurement are time-aligned. Useful for computing
+        Chi^2 stats and likelihoods.
+
+        :param measurement: measurement
+        :return: Innovation mean and covariance of hypothetical update.
+        :rtype: tuple(torch.Tensor, torch.Tensor)
+        '''
+        assert self._time == measurement.time, \
+            'State time and measurement time must be aligned!'
+
+        # Compute innovation.
+        x_pv = self._dynamic_model.mean2pv(self._mean)
+        H = measurement.jacobian(x_pv)[:, :self.dimension]
+        R = measurement.cov
+        z = measurement.mean
+        z_predicted = measurement(x_pv)
+        dz = measurement.geodesic_difference(z, z_predicted)
+        S = H.mm(self._cov).mm(H.t()) + R  # innovation cov
+
+        return dz, S
+
+    def likelihood_of_update(self, measurement):
+        '''
+        Compute and return the likelihood of a potential update, but don't
+        actually perform the update. Assumes state and measurement are time-
+        aligned. Useful for gating and calculating costs in assignment problems
+        for data association.
+
+        :param: measurement.
+        :return: Likelihood of hypothetical update.
+        '''
+        dz, S = self.innovation(measurement)
+        return torch.exp(dist.MultivariateNormal(torch.zeros_like(S), S)
+                         .log_prob(dz))
+
+    def update(self, measurement):
+        '''
+        Use measurement to update state estimate in-place and return
+        innovation. The innovation is useful, e.g., for evaluating filter
+        consistency or updating model likelihoods when the `EKFState` is part
+        of an `IMMFState`.
+
+        :param: measurement.
+        :returns: Innovation mean and covariance.
+        '''
+        assert self._time == measurement.time, \
+            'State time and measurement time must be aligned!'
+
+        x = self._mean
+        x_pv = self._dynamic_model.mean2pv(x)
+        P = self.cov
+        H = measurement.jacobian(x_pv)[:, :self.dimension]
+        R = measurement.cov
+        z = measurement.mean
+        z_predicted = measurement(x_pv)
+        dz = measurement.geodesic_difference(z, z_predicted)
+        S = H.mm(P).mm(H.t()) + R  # innovation cov
+
+        K_prefix = self._cov.mm(H.t())
+        dx = K_prefix.mm(torch.gesv(dz, S)[0])  # K*dz
```

Comment:
`gesv` returns `[d,1]` shaped tensor here that messes the shape of `_mean` later on. append `[...,0]` here and your test should work.